src/storage/src/sink/kafka.rs

```diff
@@ -1019,17 +1019,33 @@ async fn determine_sink_progress(
                 )
             })?;
 
-        progress_search(
-            &progress_client_read_committed,
-            progress_record_fetch_timeout,
-            progress_topic,
-            partition,
-            lo,
-            hi,
-            progress_key,
-            child_token.clone(),
-            Arc::clone(&metrics)
-        )
+        // This topic might be long, but the desired offset will usually be right near the end.
+        // Instead of always scanning through the entire topic, we scan through exponentially-growing
+        // suffixes of it. (Because writes are ordered, the largest progress record in any suffix,
+        // if present, is the global max.) If we find it in one of our suffixes, we've saved at least
+        // an order of magnitude of work; if we don't, we've added at most a constant factor.
+        let mut start_indices = vec![lo];
+        let mut lookback = hi.saturating_sub(lo) / 10;
+        while lookback >= 1000 {
```

Comment:
You know off the top of your head how fast we can zip through 1000 records?

Reply:
Quite fast normally - I think we saw at least tens of thousands of records per second for the slow catchups we were looking at recently. Could go higher if this seems low?

Reply:
I have now increased this to 20_000. That means that on average, for a large topic, we'll start by scanning the latest 20k to 200k records, which for the typical sink looks like it will take a couple of seconds and should include the most recent progress record even for environments with many sinks. Happy to follow up if that seems like the wrong tradeoff!